### PR TITLE
8045 backwards compatibility

### PR DESCRIPTION
**Pull request by turadg**
## Description

https://github.com/Agoric/agoric-sdk/pull/8045/ made a breaking change to Zoe, requiring contracts that used `prepare` to use `start` instead. It also updated the code for all contracts in production to use `start`, but that code is not in production. If Zoe were to be upgraded with this change, the contracts still using `prepare` would fail upon restart.

This restores support for `prepare` and sequesters the code in a region so it's easy to remove when we choose to make the breaking change. E.g. after we're confident that all vat contracts in production can accommodate it.

### Security Considerations

n/a

### Scaling Considerations

n/a

### Documentation Considerations

There's no way to strike through "prepared" in contract code because the module defines its own types. If we had [lint rules for contracts](https://github.com/Agoric/agoric-sdk/issues/4770) we could make a rule to help people migrate, but that's not a priority.

### Testing Considerations

This has a commit that updates a contract in Zoe to use `prepare` and fail. The next commit restores the compatibility so that it passes.

### Upgrade Considerations

Reduces upgrade complexity by maintaining backwards compatibility of Zoe.

8045 also moved `privateArgsShape` and `customTermsShape` to `meta`. That change is left alone as it only loosens strictness checking and doesn't affect compatibility.

### CODE REVIEW

**Comment by dckc** [comment]
> 8045 also moved privateArgsShape and customTermsShape to meta. That change is left alone as it only loosens strictness checking and doesn't affect compatibility.

Doesn't that remove protections that the API previously provided? That is: previously, if a contract exported `privateArgsShape`, it could rely on `priviateArgs` having that shape. If we stop honoring `privateArgsShape`, that guarantee goes away.

Not that contracts are in a position to defend against their creator...

**Comment by turadg** [comment]
> If we stop honoring privateArgsShape, that guarantee goes away.

Yes. But it doesn't break the execution of the contract. So it's a change to the guarantees but not a _breaking_ change.

**Comment by dckc** [comment]
> But it doesn't break the execution of the contract.

I must be missing something. Doesn't the following contract get broken?

 1. contract c1.js has `export privateArgsShape = { p1: M.nat() }`. It operates correctly if given a nat in p1, but otherwise loses customer funds.
 2. Someone tries to start c1.js using `{ p1: "hello" }`. Zoe as of mainnet 1b correctly declines to start the contract.
 3. Zoe is upgraded to ignore `export privateArgsShape ...`
 4. Someone again tries to start c1.js using ` { p1: "hello" }`. Zoe now ignores that constraint and starts the contract. The contract loses customer funds.

**Comment by turadg** [comment]
> I must be missing something.

I think we're operating with different definitions of "broken". I mean: a package has breaking changes if _merely_ adopting it causes a program depending upon it to fail. 

> Doesn't the following contract get broken?
> It operates correctly if given a nat in p1, but otherwise loses customer funds.

Only if there's an additional change in the parameters given to the contract. So not in the sense I defined above.

We could have a word for that about "change in guarantees" but I don't think that's a "build should fail" kind of change. Something for the release notes for sure. Similar to a bug fix that changes behavior.

File: packages/zoe/src/contractFacet/zcfZygote.js [hunk 1]
```diff
@@ -232,7 +238,25 @@ export const makeZCFZygote = async (
     return evalContractBundle(bundle);
   };
   // evaluate the contract (either the first version, or an upgrade)
-  const { start, buildRootObject, meta = {} } = await evaluateContract();
+  const bundleResult = await evaluateContract();
+
+  const { buildRootObject, meta = {} } = bundleResult;
+
+  //#region backwards compatibility with prepare()
```
Comment (Chris-Hibbert) [inline packages/zoe/src/contractFacet/zcfZygote.js:246]: > This ... sequesters the code in a region so it's easy to remove

I don't think the #region is serving the purpose you intend. In order to drop support for `prepare()`, different initializations for start and upgradeability would be needed.

File: packages/zoe/src/contractFacet/zcfZygote.js [hunk 2]
```diff
@@ -232,7 +238,25 @@ export const makeZCFZygote = async (
...
+  const { start: startOpt, prepare } = bundleResult;
+  // A contract must have one expression of upgradability, but it's allowed to
+  // have both the old and new way if they resolve to the same
+  if (startOpt !== prepare) {
```
Comment (Chris-Hibbert) [inline packages/zoe/src/contractFacet/zcfZygote.js]: What's the purpose of allowing `start` and `prepare` to be synonyms? I don't see how this helps contract developers.

Comment (turadg) [reply packages/zoe/src/contractFacet/zcfZygote.js]: The intent was to handle a case @dckc ran into where he wanted his contract to work in master and in this branch.

The requirement is that a developer can write a contract that works before this PR and also after. But given the conclusion (in discussion with you and @dckc ) to also be backwards compatible with the module exports, I think there's a better way to factor it. Incoming.

File: packages/zoe/test/unitTests/test-zoe-startInstance.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  const { zoe } = setup();
+
+  const contractPath = `${dirname}/beforeMetaContract.js`;
+  const bundle = await bundleSource(contractPath);
+  const installation = await E(zoe).install(bundle);
+
+  await t.throwsAsync(() => E(zoe).startInstance(installation), {
+    message: 'privateArgs: "[undefined]" - Must be: {"greeting":"hello"}',
+  });
+
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/test-zoe-startInstance.js:150]: This doesn't seem like a stable test that the contract was parsed correctly. It doesn't seem to be detecting that `prepare` without `meta` is an error.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]

File: packages/zoe/test/unitTests/test-zoe-startInstance.js (hunk 3 above)
Comment (turadg) [reply packages/zoe/test/unitTests/test-zoe-startInstance.js:150]: Prepare without meta is correct

**Review by Chris-Hibbert — APPROVED** [review]

**Pull request merged**